readme.md
# Walkner Xiconf

LED driver programming and testing application.

## Requirements

### node.js

Node.js is a server side software system designed for writing scalable
Internet applications in JavaScript.

  * __Version__: 12
  * __Website__: http://nodejs.org/
  * __Download__: http://nodejs.org/download/

## Installation

Clone the repository:

```
git clone git://github.com/morkai/walkner-xiconf.git
```

or [download](https://github.com/morkai/walkner-xiconf/zipball/master)
and extract it.

Go to the project directory and install the dependencies:

```
cd walkner-xiconf/
npm install
```

Create the `walkner-xiconf/data/` directory and give it write permissions.

## Configuration

Configuration settings can be changed by going to http://127.0.0.1:1337/#settings
(default password x!c0nf) and in the `config/frontend.js` file.

## Start

Start the application server in `development` or `production` environment:

  * under Linux:

    ```
    NODE_ENV=development node walkner-xiconf/backend/main.js ../config/frontend.js
    ```

  * under Windows:

    ```
    SET NODE_ENV=development
    node walkner-xiconf/backend/main.js ../config/frontend.js
    ```

Application should be available on a port defined in the `config/frontend.js` file
(`1337` by default). Point your Internet browser to http://127.0.0.1:1337/.

## License

This project is released under the [CC BY-NC-SA 4.0](https://raw.github.com/morkai/walkner-xiconf/master/license.md).

Copyright (c) 2016, Łukasz Walukiewicz (lukasz@miracle.systems)


history.md
2.26.7 @ 2022-01-04
===================

  * Fixed a type resulting in a NULL-reference error.

2.26.6 @ 2021-12-03
===================

  * Fixed service tag not being released if cancelled during acquisition.

2.26.5 @ 2021-09-07
===================

  * Fixed the switch action of the embedded controls.

2.26.4 @ 2021-04-02
===================

  * Updated the moment.js and moment-timezone.js.

2.26.3 @ 2021-04-02
===================

  * Changed the embedded controls to be the same as in other WMES apps.

2.26.2 @ 2021-04-01
===================

  * Changed the barcode scanning settings to only show helper barcodes relevant
    to the selected background scanner mode.

2.26.1 @ 2020-10-22
===================

  * Changed the barcode scanning to trim whitespace characters from the scanned
    value.

2.26.0 @ 2020-10-13
===================

  * Added backend scanner support on Linux.
  * Added a new scanner settings tab with scanner programming bar codes.

2.25.1 @ 2020-03-30
===================

  * Changed a font size of the order number field so its value fully fits on Linux.
  * Added two new LED board recognition patterns.

2.25.0 @ 2019-11-28
===================

  * Changed Node.js to v12 x64 and updated all dependencies.

2.24.9 @ 2019-11-05
===================

  * Fixed the MultiOne Workflow version comparison.
  * Added a new LED board recognition pattern (two scans for each board).

2.24.8 @ 2019-07-24
===================

  * Added a new LED board recognition pattern.

2.24.7 @ 2019-04-04
===================

  * Changed the UI lock to also disable all touch events.

2.24.6 @ 2019-04-04
===================

  * Added a new LED board recognition pattern.

2.24.5 @ 2018-10-24
===================

  * Added UI locking functionality.

2.24.4 @ 2018-09-27
===================

  * Changed the allowed HID ID length from 13 to up to 20.

2.24.3 @ 2018-08-31
===================

  * Changed the maximum allowed order quantity from 999 to 9999.

2.24.2 @ 2017-11-02
===================

  * Changed the default value of the Barcode scanner mode to be client-side.
  * Changed the hotkey handling to not treat a keypress as a hotkey if there
    is another keypress event in 150 ms.

2.24.1 @ 2017-08-31
===================

  * Fixed a NULL-reference error while sending a program to the GLP2 tester.
  * Changed the weight scale calibration instructions to be simpler.
  * Changed the settings import to first check the current remote server.
  * Changed the settings import to timeout each server check after 3s.

2.24.0 @ 2017-06-27
===================

  * Changed Node.js to v8.1.2 x64 and updated all dependencies.
  * Changed the default path to the MultiOne feature files.
  * Changed the single local socket check to be enabled only if the background scanner
    is enabled.
  * Added support for importing settings from the remote server.
  * Added weight scale calibration instructions.

2.23.4 @ 2017-05-24
===================
  
  * Fixed the connection to the scales not being recreated after destruction
    if there was no established connection to begin with.
  * Changed the MOW version and CoreScanner availability check to be disabled
    on non-Windows.
  * Added support for saving and reading of settings from the remote server.

2.23.3 @ 2017-03-31
===================

  * Changed the weighing barcode handling to allow spaces in the component's 12NC.

2.23.2 @ 2017-03-05
===================

  * Changed the local barcode scanner handler to include all printable characters.
    Now the scan is complete after receiving at least two characters followed by ENTER
    (to avoid conflicts with hotkeys).
  * Changed the weighing to stop the process if there is no connection to the scale.
  * Added support for min and max component weights received from the remote server.

2.23.1 @ 2017-03-05
===================

  * Added support for a new model of a scale.

2.23.0 @ 2017-02-08
===================

  * Changed the component weighing to allow multiple predefined weights
    for the same component.

2.22.1 @ 2017-01-31
===================

  * Changed the MultiOne Workflow version checking to be rerun constantly
    if the previous check failed.

2.22.0 @ 2017-01-16
===================

  * Added support for MultiOne Workflow 3.3.
  * Added support for running the app on the UP board.

2.21.0 @ 2016-11-06
===================

  * Added a new option: WEIGHT - component weight checking.

2.20.5 @ 2016-09-23
===================

  * Fixed display errors in the scanned LED board list on the history
    details page.
  * Changed the starting work mode selection process to take into
    account the currently enabled options.
  * Added a new LED board recognition pattern.

2.20.4 @ 2016-09-01
===================

  * Fixed??

2.20.3 @ 2016-09-01
===================

  * Fixed?

2.20.2 @ 2016-09-01
===================

  * Fixed a circular `require()` bug in handling GLP2 fault status when
    run in production with the require cache.
  * Changed the recently added LED board recognition pattern to be
    a little more generic.

2.20.1 @ 2016-07-13
===================

  * Added a new LED board recognition pattern.

2.20.0 @ 2016-07-06
===================

  * Changed Node.js to v6.2.2 x64 and updated all dependencies.
  * Added a new option: HID - HID lamp checking (similar to LED boards).

2.19.0 @ 2016-04-15
===================

  * Changed Node.js to v5.10.1 x64 and updated all dependencies.
  * Changed the require cache to also include the source code to speed up the app startup.
  * Changed the default remote server URL to not include the `6080` port.
  * Added a new LED board recognition pattern.
  * Added more logging messages.
  * Added an ability to download the latest logs file.

2.18.4 @ 2016-02-01
===================

  * Changed the recently added LED board matching pattern to also allow `[0-9]` where `[A-Z]` is allowed.

2.18.3 @ 2015-12-23
===================


config/frontend.js
'use strict';

var fs = require('fs');
var path = require('path');

var DATA_PATH = path.join(__dirname, '..', 'data');
var BIN_PATH = path.join(__dirname, '..', 'bin');

exports.id = 'walkner-xiconf';

exports.localSecretKey = null;

exports.modules = [
  'safeFs',
  {id: 'h5-sqlite3'},
  'settings',
  'history',
  'xiconfPrograms',
  'programmer',
  'featureSync',
  'imWorkin',
  'pubsub',
  {id: 'h5-express'},
  'httpServer',
  'sio',
  'updater'
];

exports['h5-sqlite3'] = {
  dbFile: path.join(DATA_PATH, 'db.sqlite3')
};

exports.httpServer = {
  host: '0.0.0.0',
  port: 1337
};

exports.programmer = {
  featureDbPath: path.join(DATA_PATH, 'features'),
  gprsOutputFile: path.join(DATA_PATH, 'gprs-output.xml'),
  workflowFile: path.join(DATA_PATH, 'workflow.txt'),
  lptIoFile: path.join(BIN_PATH, 'LptIo/LptIo.exe'),
  lastModeFile: path.join(DATA_PATH, 'lastMode.txt'),
  spoolFile: path.join(BIN_PATH, 'spool.exe'),
  motoBarScanFile: path.join(BIN_PATH, 'MotoBarScan/MotoBarScan.exe'),
  fakeFeatureFile: path.join(BIN_PATH, 'fake-feature.xml'),
  httpPort: exports.httpServer.port
};

exports.history = {
  featureDbPath: exports.programmer.featureDbPath,
  lastExportTimeFile: path.join(DATA_PATH, 'lastExportAt.txt')
};

exports.settings = {
  settingsFile: path.join(DATA_PATH, 'settings.json'),
  licenseEdPem: fs.existsSync(__dirname + '/license.ed.public.pem')
    ? fs.readFileSync(__dirname + '/license.ed.public.pem', 'utf8')
    : null,
  remoteServers: [],
  defaults: {
    password: 'x!c0nf',
    id: '',
    licenseKey: '',
    licenseInfo: {
      appId: null,
      appVersion: null,
      date: '0000-00-00',
      uuid: '00000000-0000-0000-0000-000000000000',
      features: 0,
      error: 'NO_KEY'
    },
    title: '',
    orders: 'disabled',
    remoteServer: '',
    syncInterval: 30,
    workflowVerify: 0,
    workflowIdentifyAlways: 0,
    workflowMultiDevice: 0,
    workflowCheckDeviceModel: 0,
    workflowCheckDevicePresent: 0,
    workflowCommissionAll: 0,
    workflowDaliFactoryNew: 0,
    featurePath1: path.join(DATA_PATH, '_features1'),
    featurePath2: path.join(DATA_PATH, '_features2'),
    searchTimeout1: 10000,
    searchTimeout2: 5000,
    readTimeout1: 5000,
    readTimeout2: 2500,
    schedulerEnabled: 0,
    schedulerFile: '',
    supportedDevicesEnabled: 0,
    supportedDevicesFile: '',
    mowLabelDataLocationEnabled: 0,
    mowLabelDataLocation: '',
    mowUniqueId: '',
    interface: 'd',
    logVerbosity: 'info',
    continueOnWarnings: 'continue',
    backupPath: 1,
    backupDelay: 15,
    cancelDelay: 0,
    programDelay: 0,
    blockageDuration: 0,
    blockageInterval: 0,
    programmerFile: 'C:/Program Files (x86)/Philips MultiOne Workflow/MultiOneWorkflow.exe',
    solComPattern: 'FTDI',
    solFilePattern: 'SOL',
    solReset: 1,
    solResetDelay: 2000,
    lptEnabled: 0,
    lptStartTimeout: 30000,
    lptFilePattern: '',
    lptReadPort: 0x378,
    lptReadBit: 15,
    lptReadInverted: 0,
    lptWritePort: 0x378,
    lptWriteBit: 5,
    imWorkin: 0,
    testingEnabled: 0,
    testingComPattern: 'COM1',
    testingComAddress: 0x00,
    testingComTimeout: 500,
    testingMaxVoltage: 24,
    testingCurrent: 10,
    testingModbusEnabled: 0,
    testingModbusHost: '127.0.0.1',
    testingModbusPort: 502,
    glp2Enabled: 0,
    glp2AllInOne: 0,
    glp2ComPattern: 'COM1',
    glp2ComAddress: 0x01,
    glp2ProgrammingDelay: 2000,
    glp2CancelDelay: 2000,
    prodLine: '',
    serviceTagPrint: 1,
    serviceTagPrinter: 'ServiceTagPrinterZPL',
    serviceTagLabelType: 'zpl',
    serviceTagLabelCode: '',
    serviceTagInLocalMode: 'optional',
    protectInputMode: 1,
    bgScanner: 0,
    bgScannerBeepBad: -1,
    bgScannerBeepGood: -1,
    bgScannerBeepDone: -1,
    bgScannerFilter: '',
    ledsEnabled: 0,
    ledsCheckingMode: 'pass',
    programming: 1,
    forceLatestOrder: 0,
    gprsVerification: 1,
    gprsVerificationTimeout: 60000,
    gprsVerificationInputPath: path.join(DATA_PATH, 'gprs', 'input'),
    gprsVerificationSuccessPath: path.join(DATA_PATH, 'gprs', 'success'),
    gprsVerificationErrorPath: path.join(DATA_PATH, 'gprs', 'error'),
    gprsOrdersPath: path.join(DATA_PATH, 'gprs', 'orders'),
    gprsProgrammerFile: 'CityTouchIPT.exe',
    gprsInputTemplateFile: path.join(DATA_PATH, 'gprs-input.json'),
    gprsDaliPort: 0,
    flResource1: '',
    flResource2: '',
    ftEnabled: false,
    ftDuration: 2000,
    ftMaxVoltage: 24,
    ftMaxResistance: 1,
    ftSetCurrent: 10,
    ftSetVoltage: 10,
    ftStartResource1: '',
    ftStartResource2: '',
    ftSerialProxyAddress: '',
    ftOrderPattern: '',
    hidEnabled: 0,
    weightEnabled: 0,
    weightHost: '',
    weightPort: 1001,
    weightStabilized: 1,
    weightTolerance: 0,
    weightCheckScope: 'none',
    weightRequireComponent: true,
    weightTimeout: 0,
    weightRefUnit: 1,
    hotkeys: {
      focusOrderNo: 'Q',
      focusQuantity: 'W',
      focusNc12: 'E',
      focusLog: 'R',
      focusHistory: 'T',
      toggleInputMode: 'A',
      toggleWorkMode: 'S',
      reset: 'D',
      reload: 'F',
      setProgram: 'G',
      printServiceTag: 'H',
      start: 'P',
      cancel: 'P',
      continue: 'L',
      showDashboardPage: 'Z',
      showHistoryPage: 'X',
      showProgramsPage: 'C',
      showSettingsPage: 'V',
      closeDialog: 'O'
    }
  }
};

exports.pubsub = {
  statsPublishInterval: 10000,
  republishTopics: [
    'featureSync.*',
    'settings.changed',
    'programmer.finished',
    'programmer.barcodeScanned',
    'programmer.startRequested',
    'programmer.resultToggled',
    'history.orderUpdated',
    'xiconfPrograms.*'
  ]
};

exports['h5-express'] = {
  mongooseId: null,
  staticPath: path.join(__dirname, '..', 'frontend'),
  staticBuildPath: path.join(__dirname, '..', 'frontend-build'),
  cookieSecret: null,
  ejsAmdHelpers: {
    t: 'app/i18n'
  },
  title: exports.id,
  routes: [
    require('../backend/routes')
  ]
};

exports.sio = {
  path: '/socket.io'
};

exports.updater = {

};
